### PR TITLE
code refactor for cors implementation fixes #103

### LINKED ISSUE #103: Fix CORS error
Currently, developers aren't able to directly use the herokuapp endpoints because CORS is denying the client requests.

Allow all domains to be able to access the public API.

### PR DESCRIPTION
**Pull request by pappyJ**
Fixes #103

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "@babel/node": "^7.10.5",
     "@babel/plugin-transform-runtime": "^7.11.5",
     "@babel/preset-env": "^7.11.5",
+    "cors": "^2.8.5",
```
Comment (ijemmao) [inline package.json:40]: It looks like your `package.json` is outdated. You should run `git pull origin master` onto your branch to get the latest updates.

Doing this is best practice to ensure that you're getting the latest changes before starting development.

**Review by ijemmao — CHANGES_REQUESTED** [review]
Please run `git pull origin master` onto your current development branch before making changes. You should always do this because you run into the situation where you're making changes to outdated files.

**Comment by ijemmao** [comment]
Also, please link to the ticket in the description of your PR

**Comment by pappyJ** [comment]
So you mean your own repo ijemmao/igbo-api ??

**Comment by ijemmao** [comment]
> So you mean your own repo ijemmao/igbo-api ??

Yeah, from my understanding your fork comes this repo so it should be in sync as much as possible with this repo. Do you see the updated files if you run `git pull origin master`?

**Comment by pappyJ** [comment]
My origin remote branch is pointing to my fork

**Comment by pappyJ** [comment]
Maybe I should add your repo to my remote branch or do I rename my origin branch to your repo?

**Comment by ijemmao** [comment]
This guide should be helpful - https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

File: tests/api-mongo.test.js [hunk 2]
```diff
@@ -1,4 +1,9 @@
 import chai from 'chai';
+<<<<<<< HEAD
```
Comment (ijemmao) [inline tests/api-mongo.test.js:2]: Seems like you have a lot of unaddressed merge conflicts.

This guide should help - https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/addressing-merge-conflicts

File: src/server.js [hunk 3]
```diff
@@ -13,6 +14,11 @@ app.use(bodyParser.urlencoded({ extended: true }));
 app.use(bodyParser.json());
 app.use(bodyParser.raw());
 
+app.use('*' , cors());
```
Comment (ijemmao) [inline src/server.js:17]: How is this line different from the `app.options('*', cors())` line? Is it possible to get rid of one these lines?

File: yarn.lock [hunk 4]
```diff
@@ -3845,7 +3845,11 @@ npm-run-path@^4.0.0:
   dependencies:
     path-key "^3.0.0"
 
+<<<<<<< HEAD
```
Comment (ijemmao) [inline yarn.lock:3848]: A quick fix for this file is just to delete and then run `yarn install` again

File: tests/api-mongo.test.js [hunk 5]
```diff
@@ -388,5 +399,30 @@ describe('MongoDB Words', function () {
           done();
         });
     });
+<<<<<<< HEAD
+});
+
+describe('API Requests For Home Directory "/"', () => {
+    describe('/undefinedRoute', () => {
+      it('should return response status of 404', (done) => {
+        chai.request('http://igboapi.com').get('/undefined').end((_, res) => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:408]: `http://igboapi.com` should be a constant, you can create the new constant inside the `tests/shared/constants.js` file.

File: tests/api-mongo.test.js [hunk 6]
```diff
@@ -388,5 +399,30 @@ describe('MongoDB Words', function () {
...
+          expect(res.status).to.equal(404);
+          done();
+        });
+      });
+    });
+  
+    it('should return hello world!', (done) => {
+      chai.request('http://igboapi.com').get('/').end((_, res) => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:416]: Again, you should make a new command inside `tests/shared/commands.js` that will handle importing the `chai` module.

Take a look at other commands like `suggestNewWord` or `suggestNewExample`.

**Review by ijemmao — CHANGES_REQUESTED** [review]
This is looking a whole lot better! There's a couple things:

* Address your merge conflicts, i.e. you should see `<<< HEAD` or `===` in your source code
* Avoid duplicating the use of `cors` in`server.js`
* A couple of notes on how to structure tests

File: tests/api-mongo.test.js (hunk 6 above)
Comment (pappyJ) [reply tests/api-mongo.test.js:416]: Noted ...

File: tests/shared/commands.js [hunk 7]
```diff
@@ -161,3 +161,9 @@ export const searchMockedTerm = (term) => {
   const regexTerm = createRegExp(term);
   return resultsFromDictionarySearch(regexTerm, term, mockedData);
 };
+
+export const getApiUrlRoute = (route = '/') => (
```
Comment (ijemmao) [inline tests/shared/commands.js:165]: nit: could you change this to `getAPIUrlRoute` so that it's consistent with `populateAPI`?

File: tests/api-mongo.test.js [hunk 8]
```diff
@@ -412,3 +416,26 @@ describe('MongoDB Words', function () {
     });
   });
 });
+
+describe('API Requests For Home Directory "/"', () => {
+  describe('/undefinedRoute', () => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:421]: I don't think you need these `describes` you can just have `the top-level `describe` and then inside of that you can have the two `it` tests

File: tests/api-mongo.test.js [hunk 9]
```diff
@@ -412,3 +416,26 @@ describe('MongoDB Words', function () {
...
+    it('should return response status of 404', (done) => {
+      const route = '/undefinedRoute';
+      getApiUrlRoute(route).end((_, res) => {
+        expect(res.status).to.equal(404);
+      });
+      done();
```
Comment (ijemmao) [inline tests/api-mongo.test.js:427]: `done` should be placed inside the `end` callback function.

Right now, you're technically calling `done` without waiting for `res` to return back to the client.

File: tests/api-mongo.test.js [hunk 10]
```diff
@@ -412,3 +416,26 @@ describe('MongoDB Words', function () {
...
+  });
+
+  describe('/', () => {
+    it('should return hello world!', (done) => {
+      getApiUrlRoute().end(async (_, res) => {
+        expect(res.status).to.equal(200);
+        expect(res.body).to.be.an('object');
+        expect(res.text).to.equal('Hello World!');
+      });
+      done();
```
Comment (ijemmao) [inline tests/api-mongo.test.js:438]: Same thing here, you want `done` to be placed inside your `end` callback.

File: tests/api-mongo.test.js [hunk 11]
```diff
@@ -412,3 +416,26 @@ describe('MongoDB Words', function () {
     });
   });
 });
+
+describe('API Requests For Home Directory "/"', () => {
+  describe('/undefinedRoute', () => {
+    it('should return response status of 404', (done) => {
+      const route = '/undefinedRoute';
+      getApiUrlRoute(route).end((_, res) => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:424]: Could you move `.end()` to its own line? Other tests do this, and it's great to make sure that they all stay consistent.

File: tests/api-mongo.test.js [hunk 12]
```diff
@@ -412,3 +416,26 @@ describe('MongoDB Words', function () {
...
+        expect(res.status).to.equal(404);
+      });
+      done();
+    });
+  });
+
+  describe('/', () => {
+    it('should return hello world!', (done) => {
+      getApiUrlRoute().end(async (_, res) => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:433]: Same thing here about moving `end` to a new line.

Also, I don't think you need to include the keyword `async`, you're not waiting for any promise to resolve inside the callback

File: tests/api-mongo.test.js (hunk 10 above)
Comment (pappyJ) [reply tests/api-mongo.test.js:438]: Placed it there but d github test failed it with some timeout stuff

File: tests/api-mongo.test.js [hunk 13]
```diff
@@ -38,6 +40,8 @@ import { expectUniqSetsOfResponses, expectArrayIsInOrder } from './shared/utils'
 const { expect } = chai;
 const { ObjectId } = mongoose.Types;
 
+chai.use(chaiHttp);
```
Comment (ijemmao) [inline tests/api-mongo.test.js:43]: You don't need to add this line since you made a new command in`commands.js` that imports `chai` and `chaiHttp`

File: tests/shared/constants.js [hunk 14]
```diff
@@ -4,6 +4,7 @@ export const LONG_TIMEOUT = 30000;
 export const EDIT_API_ROUTE = '/api/v1/edit';
 export const API_ROUTE = '/api/v1';
 export const TEST_ROUTE = '/api/v1/test';
+export const API_URL = 'http://igboapi.com';
```
Comment (ijemmao) [inline tests/shared/constants.js:7]: nice!

File: tests/api-mongo.test.js [hunk 15]
```diff
@@ -1,4 +1,5 @@
 import chai from 'chai';
+import chaiHttp from 'chai-http';
```
Comment (ijemmao) [inline tests/api-mongo.test.js:2]: Also here, no need to import this

**Comment by pappyJ** [comment]
Recap of all I need to do now pls

File: src/server.js [hunk 16]
```diff
@@ -28,6 +28,9 @@ app.use(cors({
   exposedHeaders: ['Content-Range', 'X-Content-Range'],
 }));
 
+/* implementing cors for http requests during pre - flight phase */
+app.options('*', cors());
```
Comment (ijemmao) [inline src/server.js:32]: Is it not possible to combine this logic into the already existing `cors` block above this added line? - https://github.com/ijemmao/igbo_api/blob/master/src/server.js#L27-L29

File: yarn.lock [hunk 17]
```diff
@@ -1054,6 +1054,11 @@ ansi-regex@^4.1.0:
   resolved "https://registry.yarnpkg.com/ansi-regex/-/ansi-regex-4.1.0.tgz#8b9f8f08cf1acb843756a839ca8c7e3168c51997"
   integrity sha512-1apePfXM1UOSqw0o9IiFAovVz9M5S1Dg+4TrDwfMewQ6p/rmMueb7tWZjQ1rx4Loy1ArBggoqGpfqqdI4rondg==
 
+ansi-regex@^5.0.0:
```
Comment (ijemmao) [inline yarn.lock:1057]: Since you didn't install any new packages, you shouldn't have any changes in your `yarn.lock` file. A quick and easy fix is just to delete this file and call `yarn install` again.

**Review by ijemmao — CHANGES_REQUESTED** [review]
This is getting really close! All of my comments are small fixes! Great job

File: tests/api-mongo.test.js (hunk 10 above)
Comment (ijemmao) [reply tests/api-mongo.test.js:438]: That's probably a good thing. You should investigate why making a simple `/` GET request to `http://igboapi.com/` is failing

**Comment by ijemmao** [comment]
@pappyJ What's the status with this PR? Would you like some help with this?

File: tests/api-mongo.test.js [hunk 18]
```diff
@@ -608,3 +609,24 @@ describe('MongoDB Words', () => {
...
+        done();
+      });
+  });
+
+  it('should contain Igbo API in / route', (done) => {
+    getAPIUrlRoute('/')
+      .end((_, res) => {
+        expect(res.status).to.equal(200);
+        expect(res.body).to.be.an('object');
+        expect(res.text).to.contain('Igbo API');
```
Comment (ijemmao) [inline tests/api-mongo.test.js:628]: This test is really close! But I think we want a more definitive way of checking to see if we're getting the expected page. When you hit this route you should get the homepage. I think a better check is to see if we're getting back `html` from the backend.

File: yarn.lock [hunk 19]
```diff
@@ -7982,6 +7982,8 @@ strip-ansi@6.0.0, strip-ansi@^6.0.0:
   version "6.0.0"
   resolved "https://registry.yarnpkg.com/strip-ansi/-/strip-ansi-6.0.0.tgz#0b1571dd7669ccd4f3e06e14ef1eed26225ae532"
   integrity sha512-AuvKTrTfQNYNIctbR1K/YGTR1756GycPsg7b9bdV9Duqur4gv6aKqHXah67Z8ImS7WEz5QVcOtlfW2rZEugt6w==
+  dependencies:
```
Comment (ijemmao) [inline yarn.lock:7985]: The `yarn.lock` shouldn't have changed. A quick way to fix this is to delete this `yarn.lock` and run `yarn install` again.

File: src/server.js (hunk 16 above)
Comment (ijemmao) [reply src/server.js:40]: Bumping this message. Does it make sense to have two separate instances of cors getting used in this file?

**Review by ijemmao — CHANGES_REQUESTED** [review]
This looks great! Just a couple of small changes.

**Comment by pappyJ** [comment]
What should I change ma'am?

File: src/server.js (hunk 16 above)
Comment (pappyJ) [reply src/server.js:40]: I don't know how I'd combine them .. This is how I've been doing it in my project .. Its also same with most code bases I've seen

**Comment by pappyJ** [comment]
Ma do you have any solution for d duplicates??

That's pretty as my experience level can come up with for now

On cors instance  for enabling it in all requests .. The other for special http methods and modified headers

File: src/server.js (hunk 16 above)
Comment (ijemmao) [reply src/server.js:40]: Cool, sounds like it's the norm. Just wanted to double check with you

File: tests/api-mongo.test.js [hunk 20]
```diff
@@ -608,3 +609,36 @@ describe('MongoDB Words', () => {
...
+        done();
+      });
+  });
+
+  it('/api/v1/words should get words document', (done) => {
```
Comment (ijemmao) [inline tests/api-mongo.test.js:633]: Why did you add these tests? If you look in `api-mongo.test.js` you can see that we are testing the get words route.

File: tests/api-mongo.test.js (hunk 18 above)
Comment (ijemmao) [reply tests/api-mongo.test.js:628]: @pappyJ

**Review by ijemmao — CHANGES_REQUESTED** [review]
You can click on the 'File changed' tab to see my comments in context with the code.

It looks like you're still making changes to the `yarn.lock` file. There shouldn't be able changes made to that file. I think the `git reset` command will help you in this situation.

File: tests/api-mongo.test.js (hunk 20 above)
Comment (pappyJ) [reply tests/api-mongo.test.js:633]: This one is getting data from the hosted platform

**Review by pappyJ — COMMENTED** [review]
I've seen what's going on in `yarn.lock file` ... i did everything that i could .. i even deleted d `node_modules` and yarn.lock and re installed using `yarn install` but nothing is working i also tried to pull from origin  repo but it says everything is up to date. I don,t know but since the semantic version in d project `package.json` accepts all minor changes ..could it be d cause??

**Review by ijemmao — APPROVED** [review]
